service/history/workflow/mutable_state_rebuilder.go

```diff
@@ -90,6 +92,36 @@ func NewMutableStateRebuilder(
 }
 
 func (b *MutableStateRebuilderImpl) ApplyEvents(
+	ctx context.Context,
+	namespaceID namespace.ID,
+	requestID string,
+	execution commonpb.WorkflowExecution,
+	history [][]*historypb.HistoryEvent,
+	newRunHistory []*historypb.HistoryEvent,
+) (MutableState, error) {
+	for i := 0; i < len(history)-1; i++ {
+		_, err := b.applyEvents(ctx, namespaceID, requestID, execution, history[i], nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+	newMutableState, err := b.applyEvents(ctx, namespaceID, requestID, execution, history[len(history)-1], newRunHistory)
+	if err != nil {
+		return nil, err
+	}
+	// must generate the activity timer / user timer at the very end
```

Comment:
can you entail why has to be very end?

Reply:
this is how activity & user timer works, see https://github.com/temporalio/temporal/blob/master/service/history/workflow/timer_sequence.go